presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalResourceOptimizer.java

```diff
@@ -0,0 +1,308 @@
+
+        return plan;
+    }
+
+    private int calculateHashPartitionCount(int executorCount, boolean isCurrentQueryHourPlus, boolean isRMDIRCUOptimized)
+    {
+        if (isCurrentQueryHourPlus) {
+            return (isRMDIRCUOptimized) ? calculateDircuOptimizedPartitonCountForLongerQueries(executorCount) :
+                   calculateLatencyOptimizedPartitionCountForLongerQueries(executorCount);
+        }
+        else {
+            return (isRMDIRCUOptimized) ? calculateDircuOptimizedPartitonCountForShorterQueries(executorCount) :
+                    calculateLatencyOptimizedPartitionCountForShorterQueries(executorCount);
+        }
+    }
+
+    private int calculateLatencyOptimizedPartitionCountForShorterQueries(int executorCount)
+    {
+        int lower = Math.abs(MIN_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED - executorCount);
+        int upper = Math.abs(MAX_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED - executorCount);
+
+        return ((lower - upper) > 0) ? PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_SHORTER_QUERIES :
+                PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_SHORTER_QUERIES;
+    }
+
+    private int calculateDircuOptimizedPartitonCountForShorterQueries(int executorCount)
+    {
+        int lower = Math.abs(MIN_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED - executorCount);
+        int upper = Math.abs(MAX_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED - executorCount);
+
+        return ((lower - upper) > 0) ? PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_SHORTER_QUERIES :
+                PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_SHORTER_QUERIES;
+    }
+
+    private int calculateDircuOptimizedPartitonCountForLongerQueries(int executorCount)
+    {
+        int lower = Math.abs(MIN_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED - executorCount);
+        int upper = Math.abs(MAX_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED - executorCount);
+
+        return ((lower - upper) > 0) ? PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_LARGER_QUERIES :
+                PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_LARGER_QUERIES;
+    }
+
+    private int calculateLatencyOptimizedPartitionCountForLongerQueries(int executorCount)
+    {
+        int lower = Math.abs(MIN_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED - executorCount);
+        int upper = Math.abs(MAX_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED - executorCount);
+
+        return ((lower - upper) > 0) ? PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_LARGER_QUERIES :
+                PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_LARGER_QUERIES;
+    }
+
+    private int calculateExecutorCount(double inputDataInGB, boolean isCurrentQueryHourPlus, boolean isRMDIRCUOptimized)
+    {
+        double numberOfExecutors = inputDataInGB / CONTAINER_SIZE;
```

Comment:
Container can be abstracted with cores and memory - bigger the container meaning more core and more memory more drivers can be accommodated so I would say lesser number of executors will be needed.  So the number of executors will be a function of the input data size as well as the what amount of data can be processed by a container - smaller containers with smaller core and memory will process lesser data and hence will need more executors where as for bigger containers we will need lesser number of executors. However, this is not guaranteed since vertically scaling might lead to shuffle, cosco integration issues as well in our case. So for now I am abstracting with a relationship which abstracts the containers size - we can refine more. However it makes sense to decouple verbatim from the customer siZe and have a separate tunable which we can use for different containers.